### PR TITLE
feature/1442 reduce the cognitive complexity

### PR DESCRIPTION
**Pull request by Yancy-007**
This is my second time to create the pull request. Last PR is closed by me because some problem.

**Describe the refactoring action**
An example of the issue
File link: Inception-master\inception-master\inception-search-core\src\main\java\de\tudarmstadt\ukp\inception\search\scheduling\tasks\Task.java line 139 
I reduced the two if loops to make the code look sharper and reduce the cognitive complexity.

**Expected benefit**
Cognitive complexity attempts to measure and reduce the complexity of how a method's control processes are difficult to understand. Cognitive complexity is added when the code structure that interrupts the workflow is embedded. When these parts are reconstructed, the code's comprehensibility is improved.

### CODE REVIEW

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java [hunk 1]
```diff
@@ -86,12 +87,7 @@ public boolean isReadyForPrediction(RecommenderContext aContext)
     public void train(RecommenderContext aContext, List<CAS> aCasses)
         throws RecommendationException
     {
-        List<DocumentSample> docSamples = extractSamples(aCasses);
-        
-        if (docSamples.size() < 2) {
-            LOG.info("Not enough training data: [{}] items", docSamples.size());
-            return;
-        }
+        List<DocumentSample> nameSamples = extractSamples(aCasses);
```
Comment (munterkalmsteiner) [inline inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: This change modifies the behavior of the code, i.e. is not a refactoring. Or am I wrong?

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java [hunk 2]
```diff
@@ -180,7 +176,7 @@ public EvaluationResult evaluate(List<CAS> aCasses, DataSplitter aDataSplitter)
         
         if (trainingSetSize < 2 || testSetSize < 2) {
             String info = String.format(
-                    "Not enough evaluation data: training set [%s] items, test set [%s] of total [%s].",
+                    "Not enough training data: training set [%s] items, test set [%s] of total [%s].",
```
Comment (munterkalmsteiner) [inline inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: Hm, no... why is this change necessary?

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java [hunk 3]
```diff
@@ -215,9 +211,11 @@ public EvaluationResult evaluate(List<CAS> aCasses, DataSplitter aDataSplitter)
             Type sentenceType = getType(cas, Sentence.class);
             Type tokenType = getType(cas, Token.class);
 
-            Map<AnnotationFS, List<AnnotationFS>> sentences = indexCovered(
+            Map<AnnotationFS, Collection<AnnotationFS>> sentences = indexCovered(
                     cas, sentenceType, tokenType);
-            for (Entry<AnnotationFS, List<AnnotationFS>> e : sentences.entrySet()) {
+            Collection<AnnotationFS> annotations=new ArrayList<>();//实例化的具体类型要看selectCovered(annotationType, sentence)的返回
```
Comment (munterkalmsteiner) [inline inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: If I understand the change correctly, you try here to unroll two nested loops. Since this refactoring is rather complex, I would like to see a test case that verifies that the change is not changing the behavior of the code.

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 4]
```diff
@@ -182,25 +182,23 @@ public void predict(RecommenderContext aContext, CAS aCas) throws Recommendation
             Collection<AnnotationFS> tokens = selectCovered(tokenType, sentence);
             for (AnnotationFS token : tokens) {
                 Trie<DictEntry>.Node node = aDict.getNode(text, token.getBegin());
-                if (node != null) {
+                if (node != null &&  tokens.stream().filter(t -> t.getEnd() == token.getBegin()+node.level).findAny().isPresent()) {
```
Comment (munterkalmsteiner) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Code not properly formatted (spaces missing). Please use checkstyle and eclipse code formatting style as indicated in the [developer instructions](https://inception-project.github.io//releases/0.12.2/docs/developer-guide.html#_checkstyle_and_formatting).

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java [hunk 5]
```diff
@@ -226,18 +224,18 @@ public EvaluationResult evaluate(List<CAS> aCasses, DataSplitter aDataSplitter)
                 
                 Type annotationType = getType(cas, layerName);
                 Feature feature = annotationType.getFeatureByBaseName(featureName);
+                Collections.addAll(annotations, selectCovered(annotationType, sentence));
+            }
+            for (AnnotationFS annotation : selectCovered(annotationType, sentence)) {
```
Comment (munterkalmsteiner) [inline inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: You are not using the collection `annotations` which you have populated in the first loop. Please write test cases which would find such faults.

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 6]
```diff
@@ -252,36 +250,39 @@ public EvaluationResult evaluate(List<CAS> aCasses, DataSplitter aDataSplitter)
...
+                    predictedLabel = labelStats.get(0).getLabel();
                 }
-                Optional<Span> coveringSpan = sample.getCoveringSpan(begin, end);
-                String goldLabel = NO_LABEL;
-                if (coveringSpan.isPresent()) {
-                    goldLabel = coveringSpan.get().getLabel();
-                }
-                labelPairs.add(new LabelPair(goldLabel, predictedLabel));
             }
+            Optional<Span> coveringSpan = sample.getCoveringSpan(begin, end);
```
Comment (munterkalmsteiner) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Hm, why is this now outside of the check `if (node !_ null && sample...)`? Not sure if this is correct.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/exporter/KnowledgeBaseExporter.java [hunk 7]
```diff
@@ -224,14 +224,9 @@ public void importData(ProjectImportRequest aRequest, Project aProject,
...
 
-            if (exportedKB.getRootConcepts() != null) {
-                kb.setRootConcepts(
-                    exportedKB.getRootConcepts().stream()
-                        .map(conceptId -> vf.createIRI(conceptId)).collect(Collectors.toList()));
-            }
-            else {
-                kb.setRootConcepts(new ArrayList<>());
-            }
+            kb.setRootConcepts(exportedKB.getRootConcepts() != null ? exportedKB.getRootConcepts().stream()
```
Comment (munterkalmsteiner) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/exporter/KnowledgeBaseExporter.java]: Not sure if this is cognitively easier to process...

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/render/RecommendationSpanRenderer.java [hunk 8]
```diff
@@ -188,29 +184,26 @@ public void render(CAS aCas, VDocument vdoc, AnnotatorState aState,
                 
                 boolean first = true;
                 Map<Long, AnnotationSuggestion> confidencePerClassifier = labelMap.get(label);
-                for (Long recommenderId: confidencePerClassifier.keySet()) {
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/render/RecommendationSpanRenderer.java]: I don't understand the changes that follow here... what's the purpose?

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java [hunk 9]
```diff
@@ -94,132 +94,111 @@ public void run()
             }
         };
         
-        boolean seenSuccessfulTraining = false;
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java]: This change needs definitely some verification with tests.

File: inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/scheduling/tasks/Task.java [hunk 10]
```diff
@@ -138,11 +138,10 @@ public int hashCode()
     @Override
     public boolean equals(Object obj)
     {
-        if (this == obj) {
-            return true;
-        }
-        if (obj == null) {
-            return false;
+        switch (obj){
```
Comment (munterkalmsteiner) [inline inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/scheduling/tasks/Task.java]: I don't think this works... please test.

File: inception-ui-search/src/main/java/de/tudarmstadt/ukp/inception/app/ui/search/sidebar/SearchAnnotationSidebar.java [hunk 11]
```diff
@@ -18,20 +18,22 @@
 package de.tudarmstadt.ukp.inception.app.ui.search.sidebar;
 
 import static de.tudarmstadt.ukp.clarin.webanno.api.annotation.util.WebAnnoCasUtil.getAddr;
+import static de.tudarmstadt.ukp.clarin.webanno.api.annotation.util.WebAnnoCasUtil.selectSingleFsAt;
```
Comment (munterkalmsteiner) [inline inception-ui-search/src/main/java/de/tudarmstadt/ukp/inception/app/ui/search/sidebar/SearchAnnotationSidebar.java]: Many changes here... how are they related to the refactoring?

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java (hunk 9 above)
Comment (Yancy-007) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java]: In this modification, in order to reduce the number of loops, I used a way of traversing the array to split a large loop into two small loops. This has a limited performance improvement, which greatly improves the comprehensibility of the code, so I don't think it makes sense to change the comprehensibility of the code here.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/exporter/KnowledgeBaseExporter.java (hunk 7 above)
Comment (Yancy-007) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/exporter/KnowledgeBaseExporter.java]: The changes here are a bit too deliberate. I originally wanted to reduce an if loop statement, but it actually has no effect......

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java (hunk 4 above)
Comment (Yancy-007) [reply inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Yes

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/render/RecommendationSpanRenderer.java (hunk 8 above)
Comment (Yancy-007) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/render/RecommendationSpanRenderer.java]: In order to reduce cognitive intelligibility, I want to split a large nested loop into two small loops, but the comprehensibility of the code is very unfavorable.

File: inception-ui-search/src/main/java/de/tudarmstadt/ukp/inception/app/ui/search/sidebar/SearchAnnotationSidebar.java (hunk 11 above)
Comment (Yancy-007) [reply inception-ui-search/src/main/java/de/tudarmstadt/ukp/inception/app/ui/search/sidebar/SearchAnnotationSidebar.java]: I want to split the loop as much as possible to reduce the complexity of understanding the code, but this is too mechanical, but it has a negative impact on the comprehensibility of the code.

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java (hunk 1 above)
Comment (Yancy-007) [reply inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: Yes. In order to reduce cognitive complexity, I replaced the large inline loop with two small loops. Doing so can affect functionality and is detrimental to the maintainability of the code, so this change should be removed.

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java (hunk 2 above)
Comment (Yancy-007) [reply inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: This is an unnecessary change... I originally wanted to split the if statement, but I found that it didn't make sense.

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java (hunk 5 above)
Comment (Yancy-007) [reply inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java]: There is an error in this loop, I want to reduce the loop statement but it will cause problems.

**Comment by munterkalmsteiner** [comment]
@Yancy-007 please add me as a reviewer when you have made changes and want me to check again.

**Comment by munterkalmsteiner** [comment]
@Yancy-007 please add me as reviewer when you think I should review the code again.

**Comment by reckart** [comment]
Assuming this is no longer going to be worked on.
